src/framework/emqttb_autorate.erl

```diff
@@ -16,30 +16,156 @@
 %%================================================================================
 
+start_link(Conf = #{id := Id}) ->
+  gen_server:start_link({local, Id}, ?MODULE, Conf, []).
+
 %%================================================================================
-%% Internal exports
+%% behavior callbacks
 %%================================================================================
 
+-record(s,
+        { id        :: atom()
+        , parent    :: reference() | undefined
+        , current   :: float()
+        , conf_root :: atom()
+        , error     :: fun(() -> number())
+        , last_t    :: integer()
+        , last_err  :: number()
+        }).
+
+init(Config = #{id := Id, conf_root := ConfRoot, error := ErrF}) ->
+  MRef = case Config of
+           #{parent := Parent} -> monitor(process, Parent);
+           _                   -> undefined
+         end,
+  %% Create metrics:
+  emqttb_metrics:new_gauge(?AUTORATE_RATE(Id),
+                           [ {help, <<"Controlled value">>}
+                           , {labels, [id]}
+                           ]),
+  [emqttb_metrics:new_gauge(?AUTORATE_CONTROL(Id, I),
+                            [ {help, <<"Control output delta">>}
+                            , {labels, [id, term]}
+                            ]) || I <- [sum, p, i]],
+  %% Init PID:
+  Min = my_cfg(ConfRoot, [min]),
+  Current = maps:get(init_val, Config, Min),
+  Err = ErrF(),
+  set_timer(),
+  {ok, #s{ id        = Id
+         , parent    = MRef
+         , current   = Current
+         , conf_root = ConfRoot
+         , error     = ErrF
+         , last_err  = Err
+         , last_t    = os:system_time(millisecond)
+         }}.
+
+handle_call(get_counter, _From, S) ->
+  {reply, emqttb_metrics:gauge_ref(?AUTORATE_RATE(S#s.id)), S};
+handle_call(_, _From, S) ->
+  {reply, {error, unknown_call}, S}.
+
+handle_cast(_, S) ->
+  {noreply, S}.
+
+handle_info(tick, S) ->
+  set_timer(),
+  {noreply, update_rate(S)};
+handle_info({'DOWN', MRef, _, _, _}, S = #s{parent = MRef}) ->
+  {stop, normal, S};
+handle_info(_, S) ->
+  {noreply, S}.
+
 %%================================================================================
 %% Internal functions
 %%================================================================================
+
+
+update_rate(S = #s{ last_t    = LastT
+                  , error     = ErrF
+                  , current   = CO0
+                  , last_err  = LastErr
+                  , conf_root = ConfRoot
+                  , id        = Id
+                  }) ->
+  %% 0. Get current error
+  T = os:system_time(millisecond),
+  Dt = (T - LastT) / timer:seconds(1),
+  Err = ErrF(),
+  %% 1. Get config
+  Min = my_cfg(ConfRoot, [min]),
+  Max = my_cfg(ConfRoot, [max]),
+  MaxSpeed = my_cfg(ConfRoot, [speed]),
+  Kp = my_cfg(ConfRoot, [k_p]),
+  Ti = my_cfg(ConfRoot, [t_i]),
+  %% 2. Calculate the control output
+  De = LastErr - Err,
+  CP = Kp * De, %% Note: delta t is reduced here
```

Comment:
Is this the proportional term?  If so, shouldn't it be `Kp * Err`?
If it's the derivative term, shouldn't it be divided by the time delta for the "derivative"?

Reply:
Ineed, I use "velocity form" of PID controller, that produces a derivative of the control output, rather than the absolute value. (Mostly because I was too lazy to implement a proper techinique to fight integral windup).
Since CO = DeltaC * Dt = (Kp * De / Dt + CI) * Dt = Kp * De + CI * Dt.